consensus/polybft/state.go

```diff
@@ -398,34 +398,82 @@ func (s *State) getExitEventsForProof(epoch, checkpointBlock uint64) ([]*ExitEve
 	})
 }
 
-// getExitEvents returns exit events for given epoch and provided filter
-func (s *State) getExitEvents(epoch uint64, filter func(exitEvent *ExitEvent) bool) ([]*ExitEvent, error) {
-	var events []*ExitEvent
+// updateExitEvents updates the exit events that happened in given block and old epoch to the provided new epoch
+// that means their key in db changes, so we have to remove the old event and reinsert the new one
+func (s *State) updateExitEvents(blockNumber, oldEpoch, newEpoch uint64) error {
+	return s.db.Update(func(tx *bolt.Tx) error {
+		exitEvents, err := getExitEvents(tx, oldEpoch, func(exitEvent *ExitEvent) bool {
+			return exitEvent.BlockNumber == blockNumber
+		})
 
-	err := s.db.View(func(tx *bolt.Tx) error {
-		c := tx.Bucket(exitEventsBucket).Cursor()
-		prefix := itob(epoch)
+		if err != nil {
+			return err
+		}
 
-		for k, v := c.Seek(prefix); bytes.HasPrefix(k, prefix); k, v = c.Next() {
-			var event *ExitEvent
-			if err := json.Unmarshal(v, &event); err != nil {
+		bucket := tx.Bucket(exitEventsBucket)
+
+		for _, e := range exitEvents {
+			oldKey := bytes.Join([][]byte{itob(e.EpochNumber),
+				itob(e.ID), itob(e.BlockNumber)}, nil)
+
+			if err := bucket.Delete(oldKey); err != nil {
+				// remove old event
 				return err
 			}
 
-			if filter(event) {
-				events = append(events, event)
+			e.EpochNumber = newEpoch
+			if err := insertExitEventToBucket(bucket, e); err != nil {
+				// insert updated event because its key changed
+				return err
 			}
 		}
 
 		return nil
 	})
+}
+
+// getExitEvents returns exit events for given epoch and provided filter
+func (s *State) getExitEvents(epoch uint64, filter func(exitEvent *ExitEvent) bool) ([]*ExitEvent, error) {
+	var events []*ExitEvent
+
+	err := s.db.View(func(tx *bolt.Tx) error {
+		exitEvents, err := getExitEvents(tx, epoch, filter)
+
+		if err != nil {
+			return err
+		}
+
+		events = exitEvents
+
+		return nil
+	})
```

Comment:
Small simplification proposal (using single reference for `exitEvents`, instead of two), if you find it correct:

```suggestion
	var (
		exitEvents []*ExitEvent
		err        error
	)

	err = s.db.View(func(tx *bolt.Tx) error {
		exitEvents, err = getExitEvents(tx, epoch, filter)
		if err != nil {
			return err
		}

		return nil
	})
```

Reply:
FIx: https://github.com/0xPolygon/polygon-edge/pull/1113/commits/203d13118b19e21ebdbe4b33216c1fc0170d6a7c